src/util-ebpf.c

```diff
@@ -313,141 +497,261 @@ int EBPFSetupXDP(const char *iface, int fd, uint8_t flags)
+        if (eb) {
+            eb->key[1] = key;
+        }
     }
-    return 0;
+    FLOWLOCK_UNLOCK(f);
+    return false;
+}
+
+void EBPFBypassFree(void *data)
+{
+    EBPFBypassData *eb = (EBPFBypassData *)data;
+    if (eb == NULL)
+        return;
+    if (eb->key[0]) {
+        SCFree(eb->key[0]);
+    }
+    if (eb->key[1]) {
+        SCFree(eb->key[1]);
+    }
+    SCFree(eb);
+    return;
 }
 
 /**
- * Decide if an IPV6 flow needs to be timeouted
- *
- * The filter is maintaining for each half flow a struct pair:: structure in
- * the kernel where it does accounting and timestamp update. So by comparing
- * the current timestamp to the timestamp in the struct pair we can know that
- * no packet have been seen on a half flow since a certain delay.
  *
- * If a per-CPU array map is used, this function has only a per-CPU view so
- * the flow will be deleted from the table if EBPFBypassedFlowV4Timeout() return
- * 1 for all CPUs.
+ * Compare eBPF half flow to Flow
  *
- * \param fd the file descriptor of the flow table map
- * \param key the key of the element
- * \param value the value of the element in the hash
- * \param curtime the current time
- * \return 1 if timeouted 0 if not
+ * \return true if entries have activity, false if not
  */
-static int EBPFBypassedFlowV6Timeout(int fd, struct flowv6_keys *key,
-                                     struct pair *value, struct timespec *curtime)
+
+static bool EBPFBypassCheckHalfFlow(Flow *f, EBPFBypassData *eb, void *key,
+                                    int index)
 {
-    SCLogDebug("Got curtime %" PRIu64 " and value %" PRIu64 " (sp:%d, dp:%d)",
-               curtime->tv_sec, value->time / 1000000000,
-               key->port16[0], key->port16[1]
-              );
-
-    if (curtime->tv_sec - value->time / 1000000000 > BYPASSED_FLOW_TIMEOUT) {
-        SCLogDebug("Got no packet for %d -> %d at %" PRIu64,
-                   key->port16[0], key->port16[1], value->time);
-        return 1;
+    int i;
+    uint64_t pkts_cnt = 0;
+    uint64_t bytes_cnt = 0;
+    /* We use a per CPU structure so we will get a array of values. But if nr_cpus
+     * is 1 then we have a global hash. */
+    BPF_DECLARE_PERCPU(struct pair, values_array, eb->cpus_count);
+    memset(values_array, 0, sizeof(values_array));
+    int res = bpf_map_lookup_elem(eb->mapfd, key, values_array);
+    if (res < 0) {
+        SCLogDebug("errno: (%d) %s", errno, strerror(errno));
+        return false;
     }
-    return 0;
+    for (i = 0; i < eb->cpus_count; i++) {
+        /* let's start accumulating value so we can compute the counters */
+        SCLogDebug("%d: Adding pkts %lu bytes %lu", i,
+                BPF_PERCPU(values_array, i).packets,
+                BPF_PERCPU(values_array, i).bytes);
+        pkts_cnt += BPF_PERCPU(values_array, i).packets;
+        bytes_cnt += BPF_PERCPU(values_array, i).bytes;
+    }
+    FlowBypassInfo *fc = FlowGetStorageById(f, GetFlowBypassInfoID());
```

Comment:
it seems we can optimize things by moving this to ```EBPFBypassUpdate```, that would safe us one lookup. However, we could safe us both as in the flow manager we've already looked it up just before we call ```EBPFBypassUpdate```

Reply:
Indeed